File: 152-maximum-product-subarray/152-maximum-product-subarray.cpp
class Solution {
public:
    // T.C=O(n)+O(n) ~= O(n);
    // S.C =O(1);
    
    
    int maxProduct(vector<int>& nums) {
        
        int n=nums.size();
        int pro=1;
        int ans=INT_MIN;
        
        for(int i=0;i<n;i++)
        {
            pro*=nums[i];
            
            ans=max(ans,pro);
            
            if(pro==0)
            {
                pro=1;
            }
        }
        pro=1;
        
        for(int i=n-1;i>=0;i--)
        {
            pro*=nums[i];
            
            ans=max(ans,pro);
            
            if(pro==0)
            {
                pro=1;
            }
        }
       
        return ans;
    }
};

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
         int max = nums[0];
        int min = nums[0];
        int ans = nums[0];
        
        for (int i = 1; i < nums.length; i++){
            int maxProduct = Math.max(nums[i], nums[i] * max);
            int minProduct = Math.min(nums[i], nums[i] * min);
            
			//in case of negative number flips the max and min
            maxProduct = Math.max(maxProduct, nums[i] * min); 
            minProduct = Math.min(minProduct, nums[i] * max);
            
            max = maxProduct;
            min = minProduct;
            
            ans = Math.max(ans, max);
        }
        
        return ans;
    }
}